File: ios/CellularChat/Ranging/RSSIProximityFilter.swift
import Foundation

/// Local approaching/steady/receding trend over the RSSI median history.
/// Advisory text only; RSSI never yields a direction/arrow.
enum RSSITrend { case approaching, steady, receding }
enum TrendConfidence { case low, high }

/// Filters raw BLE RSSI into coarse proximity bands (PROTOCOL_V2.md §12):
/// rolling median over the last N samples plus hysteresis so the band does not
/// flap on a single noisy reading. RSSI is never shown as an exact distance and
/// never produces a direction/arrow.
final class RSSIProximityFilter {
    private let window: Int
    private let margin: Double          // hysteresis, dBm
    private var samples: [Double] = []
    private(set) var band: ProximityBand = .unknown

    // Band boundaries in dBm (a higher/less-negative value means closer).
    private let veryNearThreshold = -55.0
    private let nearThreshold = -75.0

    // Trend regression parameters. These are pinned in code on both platforms so
    // a given (RSSI, timestamp) sequence yields identical enums; no external
    // spec defines them — `shared/PROTOCOL_V2.md` §12 covers only the bands.
    static let medianHistoryCapacity = 10       // MEDIAN_HISTORY
    static let minSamplesForTrend = 4           // MIN_SAMPLES_FOR_TREND
    static let highConfMinSamples = 6           // HIGH_CONF_MIN_SAMPLES
    /// Regression is over SECONDS, so the thresholds are a physical speed, not a
    /// per-sample step: a 1 m/s walk at distance d gives ~8.7/d dB/s. Enter at
    /// 0.6 dB/s (walking within ~14 m); measured static-noise slope sd is
    /// 0.25 dB/s at 4 dB RSSI noise, 0.38 dB/s at 6 dB, so a lower entry
    /// threshold sits inside the noise floor and labels a stationary peer.
    static let enterSlopeDbPerSecond = 0.6
    /// Hysteresis: once labelled, hold the label until the slope decays to half
    /// the entry threshold. Flipping to the OPPOSITE label still needs a full
    /// entry crossing.
    static let exitSlopeDbPerSecond = 0.3
    static let minSpanSeconds = 3.0             // no trend from a shorter arc
    static let highConfMinSpanSeconds = 6.0
    /// Residual variance of the median history about the fitted line, using the
    /// unbiased (k-2) divisor. 4 dB² ≈ 2 dB residual std: above that the link is
    /// bouncing (multipath/body blocking) and the slope means nothing.
    /// Hysteresis here too — the label is hidden at LOW, so a hard threshold
    /// would blink a latched trend on and off as the variance crosses it.
    static let residualVarianceEnter = 4.0
    static let residualVarianceExit = 8.0
    /// A hole this long (backgrounding, link stall, dropped reads) means the old
    /// samples describe a different situation: start the history over rather than
    /// regress across the gap. A backwards clock step resets for the same reason.
    static let gapResetSeconds = 10.0

    private var medianHistory: [(t: Double, rssi: Double)] = []
    private(set) var trend: RSSITrend = .steady
    private(set) var trendConfidence: TrendConfidence = .low

    init(window: Int = 5, hysteresis: Double = 5) {
        self.window = window
        self.margin = hysteresis
    }

    /// Feeds one raw reading. `timestamp` is wall-clock seconds, stamped as close
    /// to the radio read as possible (a queue hop compresses the regression arc).
    @discardableResult
    func add(rssi: Double, at timestamp: TimeInterval = Date().timeIntervalSince1970) -> ProximityBand {
        if let last = medianHistory.last {
            let elapsed = timestamp - last.t
            if elapsed < 0 || elapsed > Self.gapResetSeconds { clearHistory() }
        }
        samples.append(rssi)
        if samples.count > window { samples.removeFirst(samples.count - window) }
        // The trend history uses the averaged-middle median of the CURRENT raw
        // window (pinned cross-platform), independent of how the band rounds.
        medianHistory.append((timestamp, Self.median(samples)))
        if medianHistory.count > Self.medianHistoryCapacity {
            medianHistory.removeFirst(medianHistory.count - Self.medianHistoryCapacity)
        }
        (trend, trendConfidence) = Self.regressionTrend(
            medianHistory, current: trend, confidence: trendConfidence)
        band = classify(median: Self.median(samples), current: band)
        return band
    }

    func reset() {
        clearHistory()
        band = .unknown
    }

    /// Drops the sampling history but keeps the band: a stale window must not
    /// feed either the median or the regression after a gap.
    private func clearHistory() {
        samples.removeAll()
        medianHistory.removeAll()
        trend = .steady
        trendConfidence = .low
    }

    /// Pure least-squares trend over the median history, oldest→newest, with time
    /// (seconds) as the x-axis. `current`/`confidence` supply the hysteresis state
    /// (both thresholds are directional). The exact parameters are pinned so this
    /// is cross-platform-identical.
    static func regressionTrend(
        _ history: [(t: Double, rssi: Double)],
        current: RSSITrend,
        confidence: TrendConfidence = .low
    ) -> (RSSITrend, TrendConfidence) {
        let k = history.count
        guard k >= minSamplesForTrend else { return (.steady, .low) }
        let span = history[k - 1].t - history[0].t
        guard span >= minSpanSeconds else { return (.steady, .low) }
        let meanT = history.reduce(0) { $0 + $1.t } / Double(k)
        let meanY = history.reduce(0) { $0 + $1.rssi } / Double(k)
        var stt = 0.0, sty = 0.0
        for p in history {
            let dt = p.t - meanT
            stt += dt * dt
            sty += dt * (p.rssi - meanY)
        }
        let slope = sty / stt

        // Hysteresis: hold at the exit threshold, reverse only at the entry one.
        let trend: RSSITrend
        switch current {
        case .approaching:
            trend = slope >= exitSlopeDbPerSecond ? .approaching
                : (slope <= -enterSlopeDbPerSecond ? .receding : .steady)
        case .receding:
            trend = slope <= -exitSlopeDbPerSecond ? .receding
                : (slope >= enterSlopeDbPerSecond ? .approaching : .steady)
        case .steady:
            trend = slope >= enterSlopeDbPerSecond ? .approaching
                : (slope <= -enterSlopeDbPerSecond ? .receding : .steady)
        }

        var residualSq = 0.0
        for p in history {
            let r = p.rssi - (meanY + slope * (p.t - meanT))
            residualSq += r * r
        }
        let residualVariance = residualSq / Double(max(k - 2, 1))
        let cap = confidence == .high ? residualVarianceExit : residualVarianceEnter
        let newConfidence: TrendConfidence =
            (k >= highConfMinSamples && span >= highConfMinSpanSeconds
                && residualVariance <= cap) ? .high : .low
        return (trend, newConfidence)
    }

    private func rawBand(_ median: Double) -> ProximityBand {
        if median >= veryNearThreshold { return .veryNear }
        if median >= nearThreshold { return .near }
        return .far
    }

    private func classify(median: Double, current: ProximityBand) -> ProximityBand {
        let raw = rawBand(median)
        // First real reading, or no change: accept directly.
        if current == .unknown || raw == current { return raw }
        // Otherwise require the median to clear the crossed boundary by `margin`.
        switch (current, raw) {
        case (.far, _):
            if median >= veryNearThreshold + margin { return .veryNear }
            if median >= nearThreshold + margin { return .near }
            return current
        case (.near, .veryNear):
            return median >= veryNearThreshold + margin ? .veryNear : current
        case (.near, .far):
            return median < nearThreshold - margin ? .far : current
        case (.veryNear, _):
            if median < nearThreshold - margin { return .far }
            if median < veryNearThreshold - margin { return .near }
            return current
        default:
            return raw
        }
    }

    static func median(_ xs: [Double]) -> Double {
        guard !xs.isEmpty else { return 0 }
        let s = xs.sorted()
        let n = s.count
        return n % 2 == 1 ? s[n / 2] : (s[n / 2 - 1] + s[n / 2]) / 2
    }
}

File: android/app/src/main/java/com/cellularchat/app/ranging/RssiProximityFilter.kt
package com.cellularchat.app.ranging

/**
 * Turns raw BLE RSSI into stable proximity bands (PROTOCOL_V2.md §12,
 * IMPLEMENTATION_PLAN.md Phase 4). Two stages:
 *
 * 1. A rolling median over the last [windowSize] samples removes single-sample
 *    spikes.
 * 2. Hysteresis: a band boundary must be crossed by [hysteresisDb] before the
 *    reported band changes, so a signal hovering on a threshold does not flicker.
 *
 * Output is only ever a band — never an exact distance and never an angle.
 */
class RssiProximityFilter(
    private val windowSize: Int = 5,
    private val veryNearThresholdDb: Int = -55,
    private val nearThresholdDb: Int = -75,
    private val hysteresisDb: Int = 5,
) {
    private val window = ArrayDeque<Int>()
    private var band: ProximityBand = ProximityBand.UNKNOWN

    /** Recent (timestamp seconds, window median dBm) pairs driving the trend. */
    private val medianHistory = ArrayDeque<Pair<Double, Double>>()

    var trend: RssiTrend = RssiTrend.STEADY
        private set
    var trendConfidence: TrendConfidence = TrendConfidence.LOW
        private set

    fun current(): ProximityBand = band

    fun reset() {
        clearHistory()
        band = ProximityBand.UNKNOWN
    }

    /**
     * Feeds one raw RSSI reading (dBm) and returns the stabilized band.
     * [atMillis] is wall-clock, stamped as close to the radio read as possible:
     * the trend regresses over seconds, so a queue hop compresses its arc.
     */
    fun update(rssiDb: Int, atMillis: Long = System.currentTimeMillis()): ProximityBand {
        val timestamp = atMillis / 1000.0
        medianHistory.lastOrNull()?.let { (last, _) ->
            val elapsed = timestamp - last
            if (elapsed < 0 || elapsed > GAP_RESET_SECONDS) clearHistory()
        }
        window.addLast(rssiDb)
        while (window.size > windowSize) window.removeFirst()
        val median = medianOf(window)
        band = nextBand(band, median)
        updateTrend(timestamp)
        return band
    }

    /** Drops the sampling history but keeps the band: a stale window must not
     * feed either the median or the regression after a gap. */
    private fun clearHistory() {
        window.clear()
        medianHistory.clear()
        trend = RssiTrend.STEADY
        trendConfidence = TrendConfidence.LOW
    }

    // --- approaching/receding trend with confidence ---

    /**
     * Appends the current window's median to [medianHistory] and re-runs the
     * least-squares regression. Band and trend share the one §12 median.
     */
    private fun updateTrend(timestamp: Double) {
        medianHistory.addLast(timestamp to medianOf(window))
        while (medianHistory.size > MEDIAN_HISTORY) medianHistory.removeFirst()
        recomputeTrend()
    }

    private fun recomputeTrend() {
        val k = medianHistory.size
        if (k < MIN_SAMPLES_FOR_TREND) {
            trend = RssiTrend.STEADY
            trendConfidence = TrendConfidence.LOW
            return
        }
        val points = medianHistory.toList()
        val span = points.last().first - points.first().first
        if (span < MIN_SPAN_SECONDS) {
            trend = RssiTrend.STEADY
            trendConfidence = TrendConfidence.LOW
            return
        }
        val meanT = points.sumOf { it.first } / k
        val meanY = points.sumOf { it.second } / k
        var stt = 0.0
        var sty = 0.0
        for ((t, y) in points) {
            val dt = t - meanT
            stt += dt * dt
            sty += dt * (y - meanY)
        }
        val slope = sty / stt

        // Hysteresis: hold at the exit threshold, reverse only at the entry one.
        trend = when (trend) {
            RssiTrend.APPROACHING -> when {
                slope >= EXIT_SLOPE_DB_PER_SEC -> RssiTrend.APPROACHING
                slope <= -ENTER_SLOPE_DB_PER_SEC -> RssiTrend.RECEDING
                else -> RssiTrend.STEADY
            }
            RssiTrend.RECEDING -> when {
                slope <= -EXIT_SLOPE_DB_PER_SEC -> RssiTrend.RECEDING
                slope >= ENTER_SLOPE_DB_PER_SEC -> RssiTrend.APPROACHING
                else -> RssiTrend.STEADY
            }
            RssiTrend.STEADY -> when {
                slope >= ENTER_SLOPE_DB_PER_SEC -> RssiTrend.APPROACHING
                slope <= -ENTER_SLOPE_DB_PER_SEC -> RssiTrend.RECEDING
                else -> RssiTrend.STEADY
            }
        }

        var residualSq = 0.0
        for ((t, y) in points) {
            val r = y - (meanY + slope * (t - meanT))
            residualSq += r * r
        }
        val residualVariance = residualSq / maxOf(k - 2, 1)
        val cap = if (trendConfidence == TrendConfidence.HIGH) {
            RESIDUAL_VARIANCE_EXIT
        } else {
            RESIDUAL_VARIANCE_ENTER
        }
        trendConfidence = if (
            k >= HIGH_CONF_MIN_SAMPLES &&
            span >= HIGH_CONF_MIN_SPAN_SECONDS &&
            residualVariance <= cap
        ) {
            TrendConfidence.HIGH
        } else {
            TrendConfidence.LOW
        }
    }

    /**
     * Averaged-middle median, pinned by §12 so both platforms classify a given
     * sequence identically. Used for the band AND the trend: the band used to
     * take the upper-middle Int instead, which disagreed with iOS on every
     * even-sized window (i.e. while the window is still filling).
     */
    private fun medianOf(values: Collection<Int>): Double {
        val sorted = values.sorted()
        val n = sorted.size
        return if (n % 2 == 1) sorted[n / 2].toDouble()
        else (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0
    }

    /**
     * §12 band classification. Changing bands requires the median to clear the
     * crossed boundary by [hysteresisDb]; staying put needs nothing. Pinned
     * against `shared/vectors/rssi_bands.json`, which both platforms consume —
     * this used to be a looser rule than iOS's and the two phones of one pair
     * reported different bands at the same distance.
     */
    private fun nextBand(currentBand: ProximityBand, median: Double): ProximityBand {
        val raw = rawBand(median)
        // First real reading, or no change: accept directly.
        if (currentBand == ProximityBand.UNKNOWN || raw == currentBand) return raw
        val vn = veryNearThresholdDb
        val near = nearThresholdDb
        val m = hysteresisDb
        return when {
            currentBand == ProximityBand.FAR -> when {
                median >= vn + m -> ProximityBand.VERY_NEAR
                median >= near + m -> ProximityBand.NEAR
                else -> currentBand
            }
            currentBand == ProximityBand.NEAR && raw == ProximityBand.VERY_NEAR ->
                if (median >= vn + m) ProximityBand.VERY_NEAR else currentBand
            currentBand == ProximityBand.NEAR && raw == ProximityBand.FAR ->
                if (median < near - m) ProximityBand.FAR else currentBand
            currentBand == ProximityBand.VERY_NEAR -> when {
                median < near - m -> ProximityBand.FAR
                median < vn - m -> ProximityBand.NEAR
                else -> currentBand
            }
            else -> raw
        }
    }

    private fun rawBand(median: Double): ProximityBand = when {
        median >= veryNearThresholdDb -> ProximityBand.VERY_NEAR
        median >= nearThresholdDb -> ProximityBand.NEAR
        else -> ProximityBand.FAR
    }

    companion object {
        // Trend parameters, pinned in code on both platforms so a given
        // (RSSI, timestamp) sequence yields identical enums. No external spec
        // defines them — shared/PROTOCOL_V2.md §12 covers only the bands.
        const val MEDIAN_HISTORY = 10
        const val MIN_SAMPLES_FOR_TREND = 4
        const val HIGH_CONF_MIN_SAMPLES = 6

        /**
         * Regression is over SECONDS, so the thresholds are a physical speed,
         * not a per-sample step: a 1 m/s walk at distance d gives ~8.7/d dB/s.
         * Enter at 0.6 dB/s (walking within ~14 m); measured static-noise slope
         * sd is 0.25 dB/s at 4 dB RSSI noise and 0.38 dB/s at 6 dB, so a lower
         * entry threshold sits inside the noise floor and labels a still peer.
         */
        const val ENTER_SLOPE_DB_PER_SEC = 0.6

        /** Hold a label until the slope decays to half the entry threshold;
         * reversing to the opposite label still needs a full entry crossing. */
        const val EXIT_SLOPE_DB_PER_SEC = 0.3

        const val MIN_SPAN_SECONDS = 3.0
        const val HIGH_CONF_MIN_SPAN_SECONDS = 6.0

        /** Residual variance about the fitted line, unbiased (k-2) divisor.
         * 4 dB² ≈ 2 dB residual std; above that the link is bouncing
         * (multipath/body blocking) and the slope means nothing. Hysteresis here
         * too — the label is hidden at LOW, so a hard threshold would blink a
         * latched trend on and off as the variance crosses it. */
        const val RESIDUAL_VARIANCE_ENTER = 4.0
        const val RESIDUAL_VARIANCE_EXIT = 8.0

        /** A hole this long (backgrounding, link stall, dropped reads) means the
         * old samples describe a different situation: start over rather than
         * regress across the gap. A backwards clock step resets for the same
         * reason. */
        const val GAP_RESET_SECONDS = 10.0
    }
}
